builder/builder-config/src/main/java/io/helidon/builder/config/spi/BasicConfigResolver.java

```diff
@@ -0,0 +1,222 @@
+
+    /**
+     * Tag that represents meta information about the attribute. Used in the maps for various methods herein.
+     */
+    public static final String TAG_META = "__meta";
+
+    /**
+     * Tag that represents the component type.
+     */
+    static final String TAG_COMPONENT_TYPE = "componentType";
+
+    /**
+     * Default constructor, service loader invoked.
+     *
+     * @deprecated needed for Java service loader
+     */
+    @Deprecated
+    public BasicConfigResolver() {
+    }
+
+    @Override
+    public ConfigResolver configResolver() {
+        return this;
+    }
+
+    @Override
+    public <T> Optional<T> of(ResolutionContext ctx,
+                              Map<String, Map<String, Object>> meta,
+                              ConfigResolverRequest<T> request) {
+        Config attrCfg = ctx.config().get(request.configKey());
+        if (!attrCfg.exists()) {
+            return Optional.empty();
+        }
+
+        return optionalWrappedConfig(ctx, attrCfg, meta, request);
+    }
+
+    @Override
+    public <T> Optional<Collection<T>> ofCollection(ResolutionContext ctx,
+                                                    Map<String, Map<String, Object>> meta,
+                                                    ConfigResolverRequest<T> request) {
+        Config attrCfg = ctx.config().get(request.configKey());
+        if (!attrCfg.exists()) {
+            return Optional.empty();
+        }
+
+        return (Optional<Collection<T>>) optionalWrappedConfig(ctx, attrCfg, meta, request);
+    }
+
+    @Override
+    public <K, V> Optional<Map<K, V>> ofMap(ResolutionContext ctx,
+                                            Map<String, Map<String, Object>> meta,
+                                            ConfigResolverMapRequest<K, V> request) {
+        Config attrCfg = ctx.config().get(request.configKey());
+        if (!attrCfg.exists()) {
+            return Optional.empty();
+        }
+
+        return (Optional<Map<K, V>>) optionalWrappedConfig(ctx, attrCfg, meta, request);
+    }
+
+    private <T> Optional<T> optionalWrappedConfig(ResolutionContext ctx,
+                                                  Config attrCfg,
+                                                  Map<String, Map<String, Object>> ignoredMeta,
+                                                  ConfigResolverRequest<T> request) {
+        Class<?> componentType = request.valueComponentType().orElse(null);
+        Class<?> type = request.valueType();
+        boolean isOptional = Optional.class.equals(type);
+        if (isOptional) {
+            type = request.valueComponentType().orElseThrow();
+        }
+        boolean isList = List.class.isAssignableFrom(type);
+        boolean isSet = Set.class.isAssignableFrom(type);
+        boolean isMap = Map.class.isAssignableFrom(type);
+
+        boolean isCharArray = (type.isArray() && char.class == type.getComponentType());
```

Comment:
it is possible, but prefer to keep it the way it is for clarity.

Reply:
If I should be honest, this is more confusing than clear. If the comparison is done over the char[].class, one does not require basically not thinking about it, since it is self explanatory.

Reply:
next push.